### ISSUE TITLE
gcr.io Windows image needed for test

### ISSUE THREAD

**Issue by bclau** [kubernetes/kubernetes#74134]
**What would you like to be added**:

/sig testing
/sig windows

In the the PR https://github.com/kubernetes/kubernetes/pull/72777 we're adding a test case for pulling a Windows image from dockerhub (``should be able to pull image from docker hub [WindowsOnly]``). It has been suggested that we should also add a test case for ``gcr.io``.

However, there are no known public Windows images on gcr.io, so the only option would be to upload it ourselves. Our proposal is to have it uploaded to the ``gcr.io/kubernetes-e2e-test-images`` or ``gcr.io/google-samples`` registries (and ``gcr.io/k8s-authenticated-test`` if we want to also cover the test case ``should be able to pull from private registry with secret`` using Windows images).

The proposed image manifest list to be uploaded is ``e2eteam/busybox:1.29`` built from [1], which is currently hosted on dockerhub [2] and is currently used for testing Kubernetes with Windows nodes [3]. The manifest list contains the images:

- e2eteam/busybox:1.29-1803 (for Windows Server 1803)
- e2eteam/busybox:1.29-1809 (for Windows Server 2019)

[1] https://github.com/kubernetes-sigs/windows-testing/blob/master/images/busybox/Dockerfile
[2] https://hub.docker.com/r/e2eteam/busybox
[3] https://testgrid.k8s.io/sig-windows#Conformance%20acs-engine%20on%20Azure

**Why is this needed**:

The gcr.io Windows images will be used for the Windows equivalent test case for the following test:

- ``should be able to pull image from gcr.io``

**Comment by thockin** [comment]
It is my contention that we really only want one prod repo and that
gcr.io/kubernetes-e2e-test-images should be folded into the main repo.  The
authenticated repo should be for testing authenticated pulls and nothing
else.

To that end, I propose that we start with this.

Does the image need to be named or tagged or something to indicate OS?

How about k8s.gcr./io/testing/e2e/windows/busybox:1.29-* ?  Is that too
deep?  Does it need Windows in the path or should we put it in the tag?  Or
can we have a multi-os manifest list?

Also, is busybox the right image for testing, or can we do something even
more minimal (to address the "what f there's a busybox CVE" question)?

**Comment by ixdy** [comment]
Yes, [manifest lists support specifying the os](https://docs.docker.com/registry/spec/manifest-v2-2/#manifest-list-field-descriptions) (all images to date have used `os = "linux"`), so I'd prefer that route over yet another name or tag.

I'd also prefer something other than busybox if possible.

I originally created `gcr.io/kubernetes-e2e-test-images` because there were so many images in `gcr.io/google-containers` and it was hard to tell which of them were production images, which were test images, and which had been completely abandoned. Also, only a small number of Googlers could push to this repo since it contained all the production images.

To address the first issue, I decided that all e2e test images would be sourced from the `test/images` directory. To address the second issue, I created a new project with a wider ACL (though still currently limited to Googlers).

All that said, given the automation we're building for the main GCR repo, it'd probably be fine to fold these images back in.

**Comment by thockin** [comment]
The image promoter means that a single repo should be BETTER managed than
individual repos - we will have a PR trail of who put images there and why.

**Comment by yujuhong** [comment]
We should probably broaden the discussion to non-busybox Windows test images too. (related: #73475).

> The image promoter means that a single repo should be BETTER managed than individual repos - we will have a PR trail of who put images there and why.

Do we have a lower bar for test images? If so, having a separate repo (or making it clear that the images are test only) does have clear benefits.

**Comment by thockin** [comment]
Do we want a lower bar?  I don't think we do.

**Comment by yujuhong** [comment]
Nobody wants a lower bar until they want to add/update the image quickly for a new test, or fix a test failure to unblock the release.
If we do effectively hold the same bar today, this shouldn't be a problem.

**Comment by neolit123** [comment]
> I'd also prefer something other than busybox if possible.

i agree. also we don't need schema v2. multi-os, multi-arch image here.
just (e.g.) a nanoserver image that satisfies a Windows 1803 and Windows 1809 host.
https://hub.docker.com/_/microsoft-windows-nanoserver

a nanoserver image will obviously imply Windows only.

for a test called "should be able to pull image from gcr.io [WindowsOnly]":
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/common/runtime.go#L276-L280

all we need is to upload this image to gcr.io/[x]
and i guess we need to decide on `[x]` if we want this test to be added / passing for Windows Containers GA.

**Comment by bclau** [comment]
> 
> 
> > I'd also prefer something other than busybox if possible.
> 
> i agree. also we don't need schema v2. multi-os, multi-arch image here.
> just (e.g.) a nanoserver image that satisfies a Windows 1803 and Windows 1809 host.
> https://hub.docker.com/_/microsoft-windows-nanoserver
> 

I'm just curious, any reason why not busybox?

And indeed, a manifest list with nanoserver images could be an alternative (considering that the ``kubeletwin/pause`` image is built on top of that), but keep in mind that we'll have to run a different command on Windows (``ping -t localhost``) (we did try to find a command which works on both OSes, and ultimately found curl, but not all Linux images contain curl, check PR https://github.com/kubernetes/kubernetes/pull/72777).

**Comment by neolit123** [comment]
> I'm just curious, any reason why not busybox?

an official nanoserver image is a more secure download.

> And indeed, a manifest list with nanoserver images could be an alternative (considering that the `kubeletwin/pause` image is built on top of that)

do we need to build a nanoserver manifest list or are there official images that we can use?

i would like to bring a topic here related to pause. why are all the pause images for Windows that i found 1.1 GB or more? we definitely need the absolute minimal pause container uploaded in a location where users can download it and use that location in the guide. AFAIK, 100 MB is as low as we can go for a native Windows container, but that's still much better...

> but keep in mind that we'll have to run a different command on Windows (`ping -t localhost`) (we did try to find a command which works on both OSes

i brought this topic in the other PR. we don't need a command that works on both OSes.
if `ping -t localhost` is the command to use on Windows containers only, so be it.

**Comment by bclau** [comment]
> 
> 
> > I'm just curious, any reason why not busybox?
> 
> an official nanoserver image is a more secure download.
> 
> > And indeed, a manifest list with nanoserver images could be an alternative (considering that the `kubeletwin/pause` image is built on top of that)
> 
> do we need to build a nanoserver manifest list or are there official images that we can use?
> 

Well, technically, ``mcr.microsoft.com/windows/nanoserver:1809`` is a manifest list for arm and amd64 architectures, but obviously for the same OS version, so it's not what we want. But we can easily create one ourselves by running something like: ``docker manifest create --amend nanoserver:1.0 mcr.microsoft.com/windows/nanoserver:1809-amd64  mcr.microsoft.com/windows/nanoserver:1803``.

> i would like to bring a topic here related to pause. why are all the pause images for Windows that i found 1.1 GB or more? we definitely need the absolute minimal pause container uploaded in a location where users can download it and use that location in the guide. AFAIK, 100 MB is as low as we can go for a native Windows container, but that's still much better...
> 

Can you be a bit more specific about what pause images? If you are refering to the ``kubeletwin/pause`` image, that's basically a "pause" needed by and created for Kubelet ( https://github.com/Azure/aks-engine/blob/b1ffc7545cfedb759763d0a0c2bb44507cab8ff9/parts/k8s/windowskubeletfunc.ps1#L134 ). As you can see, it's basically the nanoserver image running ``ping -t localhost``. So, IMO, as long as we document what it is and how it is built, there is no need to publish it anywhere. Keep in mind that the nanoserver image and windowsservercore images are being updated regularly, so having the ``kubeletwin/pause`` image published somewhere would mean that it would have to be maintained and updated periodically, and the benefit for doing that is lower than the effort required.

And also, my ``kubeletwin/pause`` and ``mcr.microsoft.com/windows/nanoserver:1809`` images are 340MB in size, so your 1.1 GB image is a bit odd. can you try pulling the ``mcr.microsoft.com/windows/nanoserver:1809`` image? It might be out-of-date.

**Comment by neolit123** [comment]
> So, IMO, as long as we document what it is and how it is built, there is no need to publish it anywhere

we need to start planning how Windows containers can used easily on Windows nodes.
the pause image should be pre-built, small and easy to pull by deployment tools, such as kops, kubeadm.

> And also, my `kubeletwin/pause` and `mcr.microsoft.com/windows/nanoserver:1809` images are 340MB in size,

that is much better, but still a bit too much?
can we go down to ~100MB?

> so your 1.1 GB image is a bit odd

this image size was reported by a user in a k/k issue recently:
https://github.com/kubernetes/kubernetes/issues/74066#issuecomment-463990184

my assumption is that the user was building the image himself.

**Comment by michmike** [comment]
> > So, IMO, as long as we document what it is and how it is built, there is no need to publish it anywhere
> 
> we need to start planning how Windows containers can used be easily on Windows nodes.
> the pause image should be pre-built, small and easy to pull by deployment tools, such as kops, kubeadm.
> 
> > And also, my `kubeletwin/pause` and `mcr.microsoft.com/windows/nanoserver:1809` images are 340MB in size,
> 
> that is much better, but still a bit too much?
> can we go down to ~100MB?
> 
> > so your 1.1 GB image is a bit odd
> 
> this image size was reported by a user in a k/k issue recently:
> [#74066 (comment)](https://github.com/kubernetes/kubernetes/issues/74066#issuecomment-463990184)
> 
> my assumption is that the user was building the image himself.

i am ok to have that conversation, but please file a ticket @neolit123 and let's discuss that in sig-windows

**Comment by michmike** [comment]
everyone, thank you for your help and responses. trying to bring back this conversation to the original request. what would you like us to do in order for the few folks from google that have permissions to upload the windows images to be used for testing.

**Comment by thockin** [comment]
I am happy to help publish this via the k8s.gcr.io alias, which I
believe is the correct long-term path.

We need an image - preferably not one as prone to CVEs as busybox
historically is.  Hence a suggestion for "pause" or something
equivalently useless and tiny (if it is tiny).

Then, we need a path.

First, we'll be setting precedent for test images, so I need input on
that.  I propose `k8s.gcr./io/testing/e2e` as the root.  Objections
from test folks?

Second, do we want to squirrel this away as a Windwows-specific
binary, or should we not name it in any particular way, so that we can
pull the same image for all tests via a multi-os, multi-arch manifest
list (which can be done later).

E.g. `k8s.gcr./io/testing/e2e/does-nothing`  vs
`k8s.gcr./io/testing/e2e/windows/does-nothing`

How about k8s.gcr./io/testing/e2e/windows/busybox:1.29-* ?  Is that
too deep?  Does it need Windows in the path or should we put it in the
tag?

**Comment by bclau** [comment]
>
> We need an image - preferably not one as prone to CVEs as busybox
> historically is. Hence a suggestion for "pause" or something
> equivalently useless and tiny (if it is tiny). 

So, I'll sumarize the previous discussions. The proposed images are:

- The image currently named ``e2eteam/busybox:1.29``
  - image tags: ``e2eteam/busybox:1.29-1803`` and ``e2eteam/busybox:1.29-1809``.
  - based on: ``mcr.microsoft.com/windows/servercore:1803`` and  ``mcr.microsoft.com/windows/servercore:ltsc2019`` respectively.
  - image is currently used in Windows E2E tests and should already exist on the nodes, so download should be minimal.
- The ``mcr.microsoft.com/windows/nanoserver`` image.
  - a manifest list needs to be created containing the tags: ``mcr.microsoft.com/windows/nanoserver:1809`` and ``mcr.microsoft.com/windows/nanoserver:1803``
  - The image is used to tag the ``kubeletwin/pause`` image, which is used by kubelet, so it should already exist on the nodes.
  - smaller in size (~350MB).
  - the test should create pods with the command ``ping -t localhost`` instead of ``sleep``.

I'm fine with either one, although I think people will tend towards the nanoserver image, and in fairness, I think it will be the easiest one to maintain long-term. AFAIK, Microsoft updates their container images periodically, so, we should do the same for the gcr.io image. If not, the test using the gcr.io image will then have to pull the entire image (350 MB is ok, IMO).

>
>Then, we need a path. 
>
> First, we'll be setting precedent for test images, so I need input on
> that. I propose `k8s.gcr./io/testing/e2e` as the root. Objections
> from test folks?
>
> Second, do we want to squirrel this away as a Windwows-specific
> binary, or should we not name it in any particular way, so that we can
> pull the same image for all tests via a multi-os, multi-arch manifest
> list (which can be done later).

>E.g. `k8s.gcr./io/testing/e2e/does-nothing` vs `k8s.gcr./io/testing/e2e/windows/does-nothing`
>
>How about k8s.gcr./io/testing/e2e/windows/busybox:1.29-* ? Is that
> too deep? Does it need Windows in the path or should we put it in the
> tag?

Well, a manifest list can contain both Linux and Windows images, So, if the busybox image is picked, the Linux busybox image can be included in the manifest as well. Ideally, most images work on both OSes (with some exceptions, like the nanoserver or alpine images), so I don't think there's a need to put windows in the path.

And if a new root path is being added for new test images, I would like to ask that images mustn't container slashes in their names (e.g.: ``resource-consumer/controller``). Docker doesn't support images with slashes in their names (see https://github.com/kubernetes/kubernetes/issues/73489). Even if all the images on gcr.io eventually support Windows, it would still be a good idea to be able to use dockerhub for testing purposes.

**Comment by thockin** [comment]
>> We need an image - preferably not one as prone to CVEs as busybox

> The image currently named e2eteam/busybox:1.29
>
> image tags: e2eteam/busybox:1.29-1803 and e2eteam/busybox:1.29-1809.
> based on: mcr.microsoft.com/windows/servercore:1803 and mcr.microsoft.com/windows/servercore:ltsc2019 respectively.
> image is currently used in Windows E2E tests and should already exist on the nodes, so download should be minimal.

WHOAH.  That's a 2GB image.  For testing.

> The mcr.microsoft.com/windows/nanoserver image.
>
> a manifest list needs to be created containing the tags: mcr.microsoft.com/windows/nanoserver:1809 and mcr.microsoft.com/windows/nanoserver:1803
> The image is used to tag the kubeletwin/pause image, which is used by kubelet, so it should already exist on the nodes.
> smaller in size (~350MB).

I am in favor of this just based on the size (I was thinking something
like kuberneets' pause, which is 240 kB.  Over 1000 times smaller.  We
have different comprehensions of what "nano" means.

> I'm fine with either one, although I think people will tend towards the nanoserver image, and in fairness, I think it will be the easiest one to maintain long-term. AFAIK, Microsoft updates their container images periodically, so, we should do the same for the gcr.io image. If not, the test using the gcr.io image will then have to pull the entire image (350 MB is ok, IMO).

350 MB is *not* OK.  It's acceptable if we have no other choice, but
honestly we should create our own that is as small as humanly
possible.

>> Then, we need a path.
>>
>> First, we'll be setting precedent for test images, so I need input on
>> that. I propose k8s.gcr./io/testing/e2e as the root. Objections
>> from test folks?

I don't have enough context on e2e future, so I'd like some input from
someone who does.

>> Second, do we want to squirrel this away as a Windwows-specific
>> binary, or should we not name it in any particular way, so that we can
>> pull the same image for all tests via a multi-os, multi-arch manifest
>> list (which can be done later).
>>
>> E.g. k8s.gcr./io/testing/e2e/does-nothing vs k8s.gcr./io/testing/e2e/windows/does-nothing

> Well, a manifest list can contain both Linux and Windows images, So, if the busybox image is picked, the Linux busybox image can be included in the manifest as well. Ideally, most images work on both OSes (with some exceptions, like the nanoserver or alpine images), so I don't think there's a need to put windows in the path.

I agree (except for busybox, which is a known vector for contaigion).
Having a multi-os, multi-arch "serve_hostname" or something equally
dumb seems ideal.

> And if a new root path is being added for new test images, I would like to ask that images mustn't container slashes in their names (e.g.: resource-consumer/controller). Docker doesn't support images with slashes in their names (see #73489). Even if all the images on gcr.io eventually support Windows, it would still be a good idea to be able to use dockerhub for testing purposes.

Docker HUB doesn't.  Docker itself does, and GCR does.  If you're
using dockerhub for testing, you have to change
"k8s.gcr.io/e2e-test/whatever" to "e2eteam/whatever", right?  So what
does it matter if the GCR path is longer?

That said, I would be OK to have a single `testing` dir if that is
what test-team wants.  It just seems likely to get messy eventually.

**Comment by bclau** [comment]
>
>Docker HUB doesn't. Docker itself does, and GCR does. If you're using dockerhub for testing, you have to change "k8s.gcr.io/e2e-test/whatever" to "e2eteam/whatever", right? So what does it matter if the GCR path is longer?
>
> That said, I would be OK to have a single `testing` dir if that is what test-team wants. It just seems likely to get messy eventually.

Yes, dockerhub, I forgot the "hub" :)

And yes, that's what we're doing and it works for the most part. But, if you take a look here [1], basically, the root is ``e2eRegistry`` (by default: ``gcr.io/kubernetes-e2e-test-images``), so basically the image is: ``gcr.io/kubernetes-e2e-test-images/resource-consumer/controller``. But, if we set e2eRegistry to a dockerhub repo (e.g.: ``e2eteam``), then the image name is: ``e2eteam/resource-consumer/controller``, which, as you know, is not accepted by dockerhub.

[1] https://github.com/kubernetes/kubernetes/blob/master/test/utils/image/manifest.go#L133

**Comment by thockin** [comment]
I see.  I think it would be OK to ask that images don't use subdirs,
but the e2eRegistry variable could.

e.g.

// Docker hub
e2eRegistry = "e2eteam"

// GCR
e2eRegistry = "k8s.gcr.io/testing/e2e"

ResourceController  = Config{e2eRegistry, "resource-consumer-controller", "1.0"}


That seems workable, though it doesn't answer the question of the
actual path in GCR :)

**Comment by bclau** [comment]
> 
> I see. I think it would be OK to ask that images don't use subdirs,
> but the e2eRegistry variable could.
>
> e.g.
>
> // Docker hub
> e2eRegistry = "e2eteam"
>
> // GCR
> e2eRegistry = "k8s.gcr.io/testing/e2e"
>
> ResourceController = Config{e2eRegistry, "resource-consumer-controller", "1.0"}
>
>
>That seems workable, though it doesn't answer the question of the
> actual path in GCR :)

Yep, exactly right. :)

**Comment by michmike** [comment]
@bclau what are the next steps on this ticket to get the images to gcr.io?

**Comment by thockin** [comment]
I am still waiting for which image we are going to use, and for any
objections to "k8s.gcr./io/testing/e2e" as a root.

**Comment by bclau** [comment]
Well, I think we are in agreement that we should use the nanoserver images. As for the root repo, I have no objections.

**Comment by michmike** [comment]
@bclau can you please explicitly give @thockin all the images to upload in an authoritative list? i am hoping we can upload all of these by monday EOD.

**Comment by bclau** [comment]
Sure. As we've discussed previously, the images are:

- mcr.microsoft.com/windows/nanoserver:1809
- mcr.microsoft.com/windows/nanoserver:1803

**Comment by neolit123** [comment]
@bclau 

no manifest list image that supports both 1809 and 1803?
are we OK for only having 1809 for that "pull from GCR" test?

**Comment by bclau** [comment]
Yeah, I meant those images should be included in a single manifest list (something like ``windows-nanoserver``), like we've previously discussed.

As for the second question, AFAIK, the GCE Windows CI is using Windows Server 1803 at the moment, so it would be nice to have the new test running on that too.

**Comment by michmike** [comment]
@thockin see above for the two images we want to push. we also need to repush this image 
* resource-consumer/controller:1.0 to avoid the "/" as resource-consumer-controller
cc: @benmoss

**Comment by neolit123** [comment]
i pulled these images on Windows:
> * mcr.microsoft.com/windows/nanoserver:1809
> * mcr.microsoft.com/windows/nanoserver:1803

size of each is roughly ~350MB.

and then saved them to TAR and uploaded them here:
https://drive.google.com/drive/folders/1bw62yQijtU5mWKG5_EHWbwFQUhSrgGXT?usp=sharing

to go lower than that we might have hack custom images that trim, for instance a bunch of DLLs that we don't need  (for e.g. a minimal pause image), however such a custom image will just require yet unique download if the users are already basing their workload containers on official images.

also for such hacks we need someone to verify if the nanoserver LICENSE allows that. i'm pretty sure it doesn't.

**Comment by michmike** [comment]
no hacks please. we need to upload the images as they officially come from microsoft. :)

**Comment by spiffxp** [comment]
/milestone v1.14
We're entering [burndown](https://github.com/kubernetes/sig-release/tree/master/releases/release-1.14#burndown) and this looks relevant to the 1.14 release.  Please `/milestone clear` if I am incorrect

**Comment by benmoss** [comment]
The slash thing is kinda irrelevant to me, it's just a limitation of Docker Hub as @thockin pointed out

**Comment by michmike** [comment]
@benmoss  i understand that. but can you make forward progress in testing that autoscale ticket without this work?

**Comment by michmike** [comment]
@thockin i think you have the two images to upload. let us know when this is complete and thank you so much!

**Comment by PatrickLang** [comment]
The tars shouldn't be necessary. The nanoserver images include 2 layers that are type `application/vnd.docker.image.rootfs.foreign.diff.tar.gzip` and should not be pushed.

https://docs.docker.com/registry/spec/manifest-v2-2/

here's an example manifest from the bazel repo. you could probably use the existing rules_docker to push nanoserver with no changes if you're building and pushing images with bazel

https://github.com/bazelbuild/rules_docker/blob/master/tests/docker/windowsservercore.1803.manifest.json

**Comment by michmike** [comment]
> i pulled these images on Windows:
> 
> > * mcr.microsoft.com/windows/nanoserver:1809
> > * mcr.microsoft.com/windows/nanoserver:1803
> 
> size of each is roughly ~350MB.
> 
> and then saved them to TAR and uploaded them here:
> https://drive.google.com/drive/folders/1bw62yQijtU5mWKG5_EHWbwFQUhSrgGXT?usp=sharing
> 
> to go lower than that we might have hack custom images that trim, for instance a bunch of DLLs that we don't need (for e.g. a minimal pause image), however such a custom image will just require yet another unique download if the users are already basing their workload containers on official images.
> 
> also for such hacks we need someone to verify if the nanoserver LICENSE allows that. i'm pretty sure it doesn't.

@neolit123  we can't follow this process due to Windows licensing. i will delete your post so that it is not followed by accident.

**Comment by PatrickLang** [comment]
just to clarify - you can reference those in manifest lists, and docker push / pull them without actually pushing the layer contents to a manifest v2 complaint registry. That's the right way to do it. That separates out user-owned files in the upper layers from those in the base layers that are always served from a Microsoft server due to licensing restrictions

**Comment by bclau** [comment]
> 
> 
> just to clarify - you can reference those in manifest lists, and docker push / pull them without actually pushing the layer contents to a manifest v2 complaint registry. That's the right way to do it. That separates out user-owned files in the upper layers from those in the base layers that are always served from a Microsoft server due to licensing restrictions

Oh ok then. Keeping in mind the fact that the nanoserver image ``mcr.microsoft.com/windows/nanoserver:1809`` is a manifest itself (for amd64 and arm) and that you cannot reference another manifest list in your manifest list, then the the *actual* images which should be added to the gcr.io manifest list are:

* mcr.microsoft.com/windows/nanoserver:1809-amd64
* mcr.microsoft.com/windows/nanoserver:1803

1803 doesn't have multiarch.

**Comment by neolit123** [comment]
> @neolit123 we can't follow this process due to Windows licensing. i will hide your post so that it is not followed by accident.

the post is still quoted in your response. best to edit that as well?

> just to clarify - you can reference those in manifest lists, and docker push / pull them without actually pushing the layer contents to a manifest v2 complaint registry. That's the right way to do it. That separates out user-owned files in the upper layers from those in the base layers that are always served from a Microsoft server due to licensing restrictions

ok, looks like the licensing of these image is slightly more complicated than expected...
i hope a tutorial is not required when pushing to GCR. :)

**Comment by thockin** [comment]
What is the status of this now?  I am THOROUGHLY confused

**Comment by neolit123** [comment]
me too ` ¯\_(ツ)_/¯`

according to @PatrickLang here https://github.com/kubernetes/kubernetes/issues/74134#issuecomment-467325251 and here https://github.com/kubernetes/kubernetes/issues/74134#issuecomment-467331511
and as far as i understood, we only have to push the JSON files that hold layer references, but not the actual layers because this violates the licensing of the nanoserver images.

**Comment by PatrickLang** [comment]
I'll post explicit steps using `docker manifest create`, etc.

**Comment by PatrickLang** [comment]
## Setup prerequisites

0. Open a PowerShell window as an administrator
1. Install Docker-CE 18.09.2
```powershell
Install-Module -Name DockerMsftProvider -Repository PSGallery -Force
Install-Package -Name docker -ProviderName DockerMsftProvider
```
2. Reboot
```powershell
Restart-Computer -Force
```


## Pulling, creating the manifest, and pushing

Here's the copy & paste steps to pull the existing images, push them to the new registry, and create a manifest. Change $destinationRepo to be something like `k8s.gcr.io/microsoft/nanoserver` or your private repo name.

```powershell
$destinationRepo="patricklang/nanoserver"
$sourceRepo="mcr.microsoft.com/windows/nanoserver"
$sourceTags="1809-amd64", "1803"
$sourceTags | ForEach-Object {
    docker pull "${sourceRepo}:$_"
    docker tag "${sourceRepo}:$_" "${destinationRepo}:$_"
    docker push "${destinationRepo}:$_"
    docker manifest create --amend "${destinationRepo}:latest" "${destinationRepo}:$_"
}
```

The pushes will be nearly instantaneous because they're manifest-only pushes. All layer data is skipped.

```
The push refers to repository [docker.io/patricklang/nanoserver]
69bc1025c101: Skipped foreign layer
bc06b22070cd: Skipped foreign layer
1803: digest: sha256:01cc42375a4adcf7215ba5073e496e9a14e9a006efb2973e6db9e755841f44e7 size: 941
```

Now, you can inspect the manifest and it will show multiple images with different os.version

```powershell
docker manifest inspect "${destinationRepo}:latest"
```

```json
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 941,
         "digest": "sha256:01cc42375a4adcf7215ba5073e496e9a14e9a006efb2973e6db9e755841f44e7",
         "platform": {
            "architecture": "amd64",
            "os": "windows",
            "os.version": "10.0.17134.590"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 942,
         "digest": "sha256:f5e1be09b9696b37677c9e57bce389292c3fb845e586fd09a5d6b7fe07f9e316",
         "platform": {
            "architecture": "amd64",
            "os": "windows",
            "os.version": "10.0.17763.316"
         }
      }
   ]
}
```

And last, push it

```
docker manifest push "${destinationRepo}:latest"
```

**Comment by PatrickLang** [comment]
Here's the example repo I made on docker.io https://cloud.docker.com/repository/docker/patricklang/nanoserver. When things are done this way, the original Windows image is still served from Microsoft as required by the license. If I were to add unique layers on top of this image - that content would be pushed to the registry.

**Comment by thockin** [comment]
```
PS C:\Users\thockin> $destinationRepo="gcr.io/k8s-image-staging/testing/e2e/windows-nanoserver"
PS C:\Users\thockin> $sourceRepo="mcr.microsoft.com/windows/nanoserver"
PS C:\Users\thockin> $sourceTags="1809-amd64", "1803"
PS C:\Users\thockin> $sourceTags | ForEach-Object {
>>     docker pull "${sourceRepo}:$_"
>>     docker tag "${sourceRepo}:$_" "${destinationRepo}:$_"
>>     docker push "${destinationRepo}:$_"
>>     docker manifest create --amend "${destinationRepo}:v1" "${destinationRepo}:$_"
>> }
1809-amd64: Pulling from windows/nanoserver
Digest: sha256:f5e1be09b9696b37677c9e57bce389292c3fb845e586fd09a5d6b7fe07f9e316
Status: Image is up to date for mcr.microsoft.com/windows/nanoserver:1809-amd64
The push refers to repository [gcr.io/k8s-image-staging/testing/e2e/windows-nanoserver]
8ced52bb5964: Skipped foreign layer
37c182b75172: Skipped foreign layer
1809-amd64: digest: sha256:f5e1be09b9696b37677c9e57bce389292c3fb845e586fd09a5d6b7fe07f9e316 size: 942
Created manifest list gcr.io/k8s-image-staging/testing/e2e/windows-nanoserver:v1
1803: Pulling from windows/nanoserver
Digest: sha256:01cc42375a4adcf7215ba5073e496e9a14e9a006efb2973e6db9e755841f44e7
Status: Image is up to date for mcr.microsoft.com/windows/nanoserver:1803
The push refers to repository [gcr.io/k8s-image-staging/testing/e2e/windows-nanoserver]
69bc1025c101: Skipped foreign layer
bc06b22070cd: Skipped foreign layer
1803: digest: sha256:01cc42375a4adcf7215ba5073e496e9a14e9a006efb2973e6db9e755841f44e7 size: 941
Created manifest list gcr.io/k8s-image-staging/testing/e2e/windows-nanoserver:v1
PS C:\Users\thockin> docker manifest inspect "${destinationRepo}:v1"
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 941,
         "digest": "sha256:01cc42375a4adcf7215ba5073e496e9a14e9a006efb2973e6db9e755841f44e7",
         "platform": {
            "architecture": "amd64",
            "os": "windows",
            "os.version": "10.0.17134.590"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 942,
         "digest": "sha256:f5e1be09b9696b37677c9e57bce389292c3fb845e586fd09a5d6b7fe07f9e316",
         "platform": {
            "architecture": "amd64",
            "os": "windows",
            "os.version": "10.0.17763.316"
         }
      }
   ]
}
PS C:\Users\thockin> docker manifest push "${destinationRepo}:v1"
sha256:350f41a51a1fb33f1b0361a4247808a4b332e151a6970306822c4fd0fe98d628
```

Pushed to my staging, will try to get the promoter running ASAP

**Comment by thockin** [comment]
hmm, it looks like our image promoter is not quite ready for multi-platform yet.  Let me try pushing to gcr.io/kubernetes-e2e-test-images

**Comment by ixdy** [comment]
@thockin yeah, you might be bumping into https://github.com/google/containerregistry/issues/128.

**Comment by thockin** [comment]
```
PS C:\Users\thockin> $destinationRepo="gcr.io/kubernetes-e2e-test-images/windows-nanoserver"
PS C:\Users\thockin> $sourceRepo="mcr.microsoft.com/windows/nanoserver"
PS C:\Users\thockin> $sourceTags="1809-amd64", "1803"
PS C:\Users\thockin> $sourceTags | ForEach-Object {
>>     docker pull "${sourceRepo}:$_"
>>     docker tag "${sourceRepo}:$_" "${destinationRepo}:$_"
>>     docker push "${destinationRepo}:$_"
>>     docker manifest create --amend "${destinationRepo}:v1" "${destinationRepo}:$_"
>> }
1809-amd64: Pulling from windows/nanoserver
Digest: sha256:f5e1be09b9696b37677c9e57bce389292c3fb845e586fd09a5d6b7fe07f9e316
Status: Image is up to date for mcr.microsoft.com/windows/nanoserver:1809-amd64
The push refers to repository [gcr.io/kubernetes-e2e-test-images/windows-nanoserver]
8ced52bb5964: Skipped foreign layer
37c182b75172: Skipped foreign layer
1809-amd64: digest: sha256:f5e1be09b9696b37677c9e57bce389292c3fb845e586fd09a5d6b7fe07f9e316 size: 942
Created manifest list gcr.io/kubernetes-e2e-test-images/windows-nanoserver:v1
1803: Pulling from windows/nanoserver
Digest: sha256:01cc42375a4adcf7215ba5073e496e9a14e9a006efb2973e6db9e755841f44e7
Status: Image is up to date for mcr.microsoft.com/windows/nanoserver:1803
The push refers to repository [gcr.io/kubernetes-e2e-test-images/windows-nanoserver]
69bc1025c101: Skipped foreign layer
bc06b22070cd: Skipped foreign layer
1803: digest: sha256:01cc42375a4adcf7215ba5073e496e9a14e9a006efb2973e6db9e755841f44e7 size: 941
Created manifest list gcr.io/kubernetes-e2e-test-images/windows-nanoserver:v1
PS C:\Users\thockin>
PS C:\Users\thockin> docker manifest inspect "${destinationRepo}:v1"
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 941,
         "digest": "sha256:01cc42375a4adcf7215ba5073e496e9a14e9a006efb2973e6db9e755841f44e7",
         "platform": {
            "architecture": "amd64",
            "os": "windows",
            "os.version": "10.0.17134.590"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 942,
         "digest": "sha256:f5e1be09b9696b37677c9e57bce389292c3fb845e586fd09a5d6b7fe07f9e316",
         "platform": {
            "architecture": "amd64",
            "os": "windows",
            "os.version": "10.0.17763.316"
         }
      }
   ]
}
PS C:\Users\thockin> docker manifest push "${destinationRepo}:v1"
sha256:350f41a51a1fb33f1b0361a4247808a4b332e151a6970306822c4fd0fe98d628
```

Someone test that pull?

**Comment by thockin** [comment]
@ixdy does that mean no Windows images can live on GCR?  Am I misreading it?

**Comment by ixdy** [comment]
@thockin I believe GCR supports it, but some of our tooling (which uses google/containerregistry) may not.

OTOH I think @listx's tool is just using `gcloud` directly, so it might be unaffected.

**Comment by neolit123** [comment]
> Someone test that pull?

![nano-server-image-gcr](https://user-images.githubusercontent.com/331852/53454362-92aa0280-3a2f-11e9-864b-022eba1a5a52.png)

**Comment by thockin** [comment]
Awesome, so that seems to work for now.  I guess the next step is to FROM that in your own containers.

**Comment by thockin** [comment]
or just use it in your tests.  Do we close this for now, then?

**Comment by neolit123** [comment]
SGTM, leaving the close call to @PatrickLang @bclau

**Comment by PatrickLang** [comment]
This looks right. I did a pull on a 1809 machine and the hashes match.

```
gcr.io/kubernetes-e2e-test-images/windows-nanoserver   v1                                         d65434b3ecb4        2 weeks ago         347MB
mcr.microsoft.com/windows/nanoserver                   1809-amd64                                 d65434b3ecb4        2 weeks ago         347MB
```

@bclau is this enough to finish that test case, or are more images needed?

We definitely need to get the images into a CI for 1.15. See #73475 / https://github.com/kubernetes-sigs/windows-testing/issues/3

**Comment by neolit123** [comment]
> is this enough to finish that test case, or are more images needed?

this should be enough in terms of the test case.

**Comment by neolit123** [comment]
@yujuhong

> gcr.io/kubernetes-e2e-test-images/windows-nanoserver:v1

^ FYI, there are now Windows test images at GCR.

thanks @thockin !

**Comment by bclau** [comment]
> 
> 
> This looks right. I did a pull on a 1809 machine and the hashes match.
> 
> ```
> gcr.io/kubernetes-e2e-test-images/windows-nanoserver   v1                                         d65434b3ecb4        2 weeks ago         347MB
> mcr.microsoft.com/windows/nanoserver                   1809-amd64                                 d65434b3ecb4        2 weeks ago         347MB
> ```
> 
> @bclau is this enough to finish that test case, or are more images needed?
> 
> We definitely need to get the images into a CI for 1.15. See #73475 / [kubernetes-sigs/windows-testing#3](https://github.com/kubernetes-sigs/windows-testing/issues/3)

Awesomeness, I've made a PR adding the testcase using this new image. This issue will close automatically once that merges. Test works for me.

One question remains though. Do we have to have a Windows equivalent for the ``should be able to pull from private registry with secret`` test too? The test is currently pulling the image ``gcr.io/authenticated-image-pulling/alpine:3.7``, so it fails on Windows.

If we have to add a Windows equivalent for that too, then the nanoserver manifest list should also be pushed to the ``gcr.io/authenticated-image-pulling`` registry as well.

Anyways. thanks for the help, everyone!

**Comment by michmike** [comment]
@bclau  are you asking if @thockin  can push the nanoserver manifest to the authenticated-image-pulling registry as well? Tim?

**Comment by bclau** [comment]
> 
> 
> @bclau are you asking if @thockin can push the nanoserver manifest to the authenticated-image-pulling registry as well? Tim?

Yep. It is a conformance test, so it would be nice to be able to cover it as well.

**Comment by michmike** [comment]
@thockin  any chance you push the nanoserver manifest to the authenticated-image-pulling registry as well?

thanks for all your help! once you do the above, we can lose this issue

**Comment by thockin** [comment]
Try "gcr.io/authenticated-image-pulling/windows-nanoserver:v1"

**Comment by neolit123** [comment]
thanks it works!
![docker-pull-auth](https://user-images.githubusercontent.com/331852/53606943-63c59500-3bc5-11e9-80b3-a89f5759664f.png)

**Comment by neolit123** [comment]
@bclau please, close the issue once you check this.

**Comment by michmike** [comment]
thanks Tim! you rock!

**Comment by michmike** [comment]
@bclau  can you please close this issue if we are done?

**Comment by bclau** [comment]
The issue is supposed to close when the PR https://github.com/kubernetes/kubernetes/pull/74655 merges. Closing anyways.
